### DISCUSSION TITLE
V2.0.0.x Alpha 1 Discussion

### DISCUSSION THREAD

**Discussion by AdamC1228** [Erabior/RouteManager#57] [General]
Hello Everyone,

I present the changes / improvements for V2.0.0.x Alpha 1 for discussion.

**This is Alpha software: An early version of a program or application, typically unstable, but useful to show what the product will do. Sometimes this stage is referred to as a preview version.** 

**Improvements:**
Highlighted below are the overall changes between V1 & V2. They include but are not limited to:

**Important Stuff**
- UI Updates
    - Increased the car inspector window size 
		- Scroll-able list is now usable and shows more than a single option at a time.
	- Implemented feature request #30 
		- You can now define the maximum speed of the locomotive based on original speed slider.
		- Original speed slider now added back to the game
		- ![image](https://github.com/Erabior/RouteManager/assets/13957003/389328a7-830b-4286-9b93-9c1135409f4f)
- Scratch rewrite of the passenger loading and unloading logic
    - Passenger loading now scans the passengers at the station and will prevent departure if any passengers on the platform are destined for a selected station.
	- Passenger unloading now scans the entire consist and identifies if any coaches contain passengers destined for the current station and prevents departure until no passengers remain for the current station stop.
- Improved direction controls
    - Better tracking of locomotive direction so that while enroute, locomotive should not start then stop and change directions.
        - Locomotive may still do an initial start stop and reverse, upon starting the route but otherwise the frequency should be reduced greatly.
        - Locomotive should retain its sense of direction after decoupling for fueling and re-coupling to the train. 

**Cool Stuff**
- Single station operation mode improvements
    - Single station operation has been improved
	    - Theoretical support for Freight only operations thus allowing you to tell a freight train to drive to station X where upon it will stop.

**Slightly Boring Stuff**
- Improved station selection / determination logic
    - Rewrote portions of the station selection / destination logic to aide in the above improvements
- Improved station Approach and stopping logic
    - Implemented stepped logic for Long Approach, Medium Approach, Short Approach, and On Arrival 
        - Each approach slowly reduces the maximum allowable speed to mimic a more gradual application of brakes when entering a station
        - Also reduces chances of locomotives sometimes stopping abruptly and or stopping short and having to restart to finish entering the platform
- Bell application Improvements
    - Bell now activates if the train is on approach to a station and either
	    - Enters the platform
		- Speed drops to 5 mph
		
**Mostly Boring Stuff**
- Refactoring
    - Rewrote large portions of the mod
	    - Split several large chunks of code into several smaller more manageable methods with an identifiable purpose
		- Separated the mod further into more sub-files and classes with common purposes 
	- Comments are now prevalent throughout the majority of the code base.
- User adjustable settings referenced from an INI config file (See Dev branch read-me for more information)
    - Log Levels are adjustable from the new customization file
    - Dev features can be toggled on and off
    - Quantities used for the fuel alerts are definable based on user preference
	
**Actually Boring Stuff**
- Updated Read-Me
    - Added language / sections relevant to the new customization's available via the INI file
	- Updated some minor formatting / typos
- Logging
    - Added Configurable debug levels
        - Trace, Verbose, Debug, Info, Warning, Error
		- Updated Logging to Identify the current debug level in the log file
		- Updated Logging to Identify which error level a specific message is derived from.

**Comment by Voidheartd** [comment]
Does this version handle switching to manual (for refueling etc.) and back better?

**Reply by AdamC1228** [reply]
It should but since so much code has been rewritten and this is the first Alpha release, i cannot make guarantees yet. I can only say that in my limited testing things behaved better across the board with several variations of stopping and resuming.

**Comment by Kaceyproductions** [comment]
Alright, a short amount of testing has gotten some results.

First off, setting destinations on passenger cars seems to work much better, though it seems to still behave a bit better when you do it manually.

Beyond that, the new way orders are handled is.. rough. 

I was in Dillsboro, told the AI to go to Sylva. Intended on setting destinations on the way, so it would stop and head west after loading. It did not load any passengers, and just stayed at Sylva. It seems the orders are more or less set in stone when you first turn on route mode, meaning starting in the middle of the line is a no-go. 

Speaking of turning on route mode, clicking it off and back on to "refresh" seems to break the speed setting. It will just stay on whatever setting it was on, and never change unless you do it manually. 

And then last thing I noticed, though I couldn't recreate this: The train doesn't seem to know where it's at sometimes. Departed Sylva for Dillsboro, but the AI logged Wilmot for Whittier. This also made it completely ignore Dillsboro. Only got this to happen the once, though. Subsequent testing didn't have this happen again. 

Only thing I didn't try was leaving Dillsboro with the orders set to everything west, and enabling Sylva. That *might* make the orders refresh and work right. Will do more testing later and report back

**Reply by AdamC1228** [reply]
The following are my comments and responses to the points you bring up.

- Setting destinations working better is a plus
    - After feedback and thinking on it I think I see the missing edge case to fully address this. I will address the edge case in the next Alpha.
- Settings destinations enroute not working as well as before
    - Am I correct in understanding that you are telling a train to drive to a single station. Upon arriving at that station you enable the rest of the stations that you want the consist to travel to?
    - If so, This sounds like an edge case. When operating in single station mode, the AI when it reaches its destination now understands and checks for if the station is the only station in the list and will subsequently terminate operation with the Console Message: "{LOCO} has no more stations. Halting Control. Can you please confirm if this is the case?
    - Also try the following: Set Sylva first and then enable all the remaining stations once the consist leaves the station. Under the new logic the loco will continue to proceed in whatever direction you tell it to travel regardless of whether you add additional stations in the opposite direction.
- Speed setting resetting 
    - I had addressed this already but it appears to be genuine regression. This will be fixed in V2 Alpha 2.
- Train is lost and seems brain-dead
    - If you do manage to find a way to reproduce please let me know because this has me curious.

This is why its a pre-release alpha after all ;)

**Reply by Kaceyproductions** [reply]
> Am I correct in understanding that you are telling a train to drive to a single station. Upon arriving at that station you enable the rest of the stations that you want the consist to travel to?

Not quite, no. 

> When operating in single station mode, the AI when it reaches its destination now understands and checks for if the station is the only station in the list and will subsequently terminate operation with the Console Message: "{LOCO} has no more stations. Halting Control. Can you please confirm if this is the case?

Yup, that works just fine, though it also leads into the speed resetting issue once it's done.

> Also try the following: Set Sylva first and then enable all the remaining stations once the consist leaves the station. Under the new logic the loco will continue to proceed in whatever direction you tell it to travel regardless of whether you add additional stations in the opposite direction.

This is what I was meaning originally. While underway to Sylva, I set all the other stations. It gets to Sylva, then just.. stops. No message about "no further orders" and there's no passengers to load/unload (after a few moments). It just.. parks and says nothing.

Anyway, glad to lend a hand in stress-testing such an awesome mod. Will definitely report back more findings once I get more testing done.

**Reply by AdamC1228** [reply]
Something does not quite add up here. It sounds like a but that was patched several commits ago. Can you please please set the log to trace and reproduce the no load / unload for me please. I will try to review them and do some testing on Friday. 


**Additionally, I suspect something went wrong and re-uploaded the .dll for v2 alpha 1. Please try the re-uploaded copy.**

**Comment by GamearooCodes** [comment]
will 2.0 include a option for the ai to auto refuel water and coal as it needs it as that will be nice

**Reply by AdamC1228** [reply]
I will defer to @Erabior but I don't have any qualms with considering this but I would like to get what we have already identified in alpha 1 ironed out before commiting to anything further.

**Reply by Erabior** [reply]
The thing about auto coaling/watering/refueling is the need to throw switches. While this can be done as another modder has shown me that each switch has an I'd and can be operated semi easily from code, this would have to be essentially hard coded for each case. I expect the first foray into automatic switching will be to handle the Alarka branch. This in my opinion would be significantly easier to implement. Furthermore when issuing commands to the road AI there is a slight delay which would make it difficult to properly spot the locomotive for watering, coaling, and refueling. We would essentially have to take direct control of the locomotive which I haven't looked into yet either so it would require not much but definitely a little bit of digging into the game's code. Finally actually determining when the locomotive is spotted could prove challenging. Some locomotives have rather large doors for the water tank but others are quite small and the position requirements are rather finicky. Even manually watering I have found it somewhat easy to overshoot or undershoot the water tower.

**Reply by AdamC1228** [reply]
My thoughts exactly however I am hopeful that the developers based on the video they shared heading into the holidays will have done most of the heavy lifting for us when they release their update to route mode. The teaser showed that you can program a destination and the game will "route" the switches accordingly. We would just need to integrate with that logic where possible when they release the feature which would solve the switching part of the equation. 

As for alignment yeah I have a few initial thoughts none of them terribly great but that is the kind of problem you will only solve if you sit down and dedicate the effort to it.

**Comment by Cathal-Brugha** [comment]
Just tried Alpha v2 build. Train just takes off in whatever direction is selected before you select stations and enable route. For example, loaded save, clicked road, forward was selected, checked box for Sylva (train needs to go reverse to get to Sylva station), checked enable route, set speed 35, train took off forward to end of line. Tried again, clicked Road, selected reverse, checked box for Sylva, checked Enable Route, set speed 35, train took off in reverse and didn't stop at Sylva.

[Player.log](https://github.com/Erabior/RouteManager/files/13837489/Player.log)

**Reply by armyf35** [reply]
Are you in the same situation I was with #59, being company mode and only some milestones unlocked?

**Reply by Cathal-Brugha** [reply]
Possibly. I just unlocked Bryson to Whittier signals, and just completed the first phase of Bryson to Fontana bridge. Stable version works ok, apart from issue #54 where I need to manually set the cars up.

**Reply by AdamC1228** [reply]
Log appears to indicate issue #59. Please replicate with the log option in the config set to verbose. @Cathal-Brugha. Hopefully can be worked around short term while we wait to see how / if the devs address the underlying issue.

**Reply by Cathal-Brugha** [reply]
What config file, located where?

**Reply by AdamC1228** [reply]
The included INI file for V2. If you look at the[ Alpha branch readme](https://github.com/Erabior/RouteManager/tree/v2-Alpha) it should contain the information that you need.

**Reply by Cathal-Brugha** [reply]
Ok, here is the new log
[Player.log](https://github.com/Erabior/RouteManager/files/13850064/Player.log)

**Reply by AdamC1228** [reply]
@Cathal-Brugha confirmed to be issue #59 causing unexpected behavior.

**Reply by AdamC1228** [reply]
@Cathal-Brugha Please follow up in the following bug report with your log files: https://discord.com/channels/795878618697433097/1192341645891473438

**Comment by Voidheartd** [comment]
Were the console announcements for arriving/departing removed in this? That is a very useful feature that I'd like to see continue to be included.

**Reply by AdamC1228** [reply]
No, due to some of the catalogued known issues it is more often than not reaching that point in the code. I have not experience this in my play testing, then again I have not seen most of the raised bugs in my play-testing either. Once some of the other bugs are addressed for Alpha 2, you should start seeing them again.

**Comment by GamearooCodes** [comment]
How is the Alaskan jct handled does it notify when a switch needs alignment for the next station bc ik it will have to be switched to get to the next station if I'm not mistaken

**Reply by AdamC1228** [reply]
Currently in V2 no special attention has been given to Alarka Jct. In fact, it potentially could be worse than V1 at the current stage in development as no testing has been done with the new code base for that area yet. 

The approach thus far has been, lets get the core feature base of V2 with all the improvements & new features relatively stable before worrying about Alarka Jct.
